src/CraftingRecipes.h

```diff
@@ -132,11 +133,18 @@ class cCraftingRecipes
 	{
 		cRecipeSlots m_Ingredients;
 		cItem        m_Result;
+		AString      m_RecipeName;
```

Comment:
Minecraft allows generating certain data, like the recipes (https://wiki.vg/Data_Generators). I use the file names as `m_RecipeName` which uniquely identifies each recipe. This makes it easy to extend for other Minecraft versions.

Of course, we could use IDs instead, if you think an additional layer of abstraction makes sense, I can look into it.

Reply:
>This makes it easy to extend for other Minecraft versions.

An internal ID doesn't need to be stable across minecraft versions because we need to translate it to send over the protocol anyway. The index into `m_Recipes` would be a good unique identifier for example.

Reply:
Before starting to implement this, I would like to understand the advantage and get on the same page.

`Server/crafting.txt` (resolving the cuberite recipes to the Minecraft recipes) IMHO we should use the Minecraft names to be able to compare it easily with the official ones and be able to maintain it easily.
`Server/Protocol/x.x.x/base.recipes.txt` (mapping the recipes to the protocol IDs) IMHO we should use the Minecraft names to be able to understand the file easily and don't have to look up every line `somewhere` else to get a clue.
`Server/players/xx/xxx.json` (the list of known recipes per player) IMHO we should use the Minecraft names, which makes it easy to understand the player data. More important if we use custom cuberite IDs and these IDs change for some reason, it will be hard to trace issues in the future.

Does this make sense so far?

So looking only into the server-internal handling:

In `CraftingRecipes.h - cRecipe` the `AString m_RecipeName` could be replaced by an ID (as suggested here), but this also means that there is another mapping of the new ID to the recipe name. If the intent is saving memory, this would be the opposite.

In `RecipeMapper` the mapping to the different protocol IDs could be done with only IDs, `Cuberite specific recipe ID => Minecraft protocol recipe ID`, this could save some memory.

So internally I can understand that this could make sense. IMHO this is a tradeoff between code complexity and minimal increased server start time vs memory consumption.
If this is the current blocker, I will look into it and provide the necessary changes.

Reply:
>Does this make sense so far?

Yes. Files should use the names, that's fine.

>In `CraftingRecipes.h - cRecipe` the `AString m_RecipeName` could be replaced by an ID (as suggested here), but this also means that there is another mapping of the new ID to the recipe name. If the intent is saving memory, this would be the opposite.

I don't literally mean replace `m_RecipeName` with an ID. I mean to refer to the recipe in code by an ID rather than by name. For example, `findNewRecipesForItem` returns an array of recipe names. In doing so, we have to allocate potentially hundreds of new strings. Instead, we could just return an array of numbers requiring just one allocation.

Also, `findRecipeByRecipeName` for example goes from an O(n) linear search to being an O(1) array index operation if you go for IDs as indexes into the recipe list.

Reply:
Ok, I get it (I hope :-))

I will look into it and let you know once I adapted accordingly.

Btw. thanks for taking the time to look into it and explain it in more detail.